dali/operators/math/expressions/broadcasting.h

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_MATH_EXPRESSIONS_BROADCASTING_H_
+#define DALI_OPERATORS_MATH_EXPRESSIONS_BROADCASTING_H_
+
+#include <utility>
+#include "dali/core/tensor_shape.h"
+#include "dali/core/tensor_shape_print.h"
+
+namespace dali {
+
+/**
+ * @brief Returns the number of dimensions of the broadcasted shape
+ */
+DLL_PUBLIC int BroadcastNdim(span<const TensorShape<>> shapes);
+DLL_PUBLIC int BroadcastNdim(span<const TensorListShape<>> shapes);
+
+/**
+ * @brief Verifies that all shapes have the same number of samples
+ */
+DLL_PUBLIC void CheckNumSamples(span<const TensorListShape<>> shapes);
+
+/**
+ * @brief Calculates the resulting shape of broadcasting two or more compatible shapes.
+ *        It is allowed that the shape has different number of dimensions, in which case 
+ *        the shape with fewer dimensions will be prepended with leading dimensions
+ *        with extend 1.
+ *        Example: lhs=(10, 10, 3), rhs=(1, 3) -> result=(10, 10, 3)
+ * 
+ * @param result resulting shape (should as many dimensions as the operands)
+ * @param lhs shape of left hand side operand
+ * @param rhs shape of right hand side operand
+ * @return TensorShape<> resulting shape
+ */
+DLL_PUBLIC void BroadcastShape(TensorShape<>& result, span<const TensorShape<>> shapes);
+DLL_PUBLIC void BroadcastShape(TensorListShape<>& result, span<const TensorListShape<>> shapes);
+
+template <typename Shape>
+void BroadcastShape(Shape &result, const Shape& a, const Shape& b) {
```

Comment:
This overload might be quite expensive, if we decide to care about it, as it creates a copy of the inputs, same below, for other overloads of this kind.

Reply:
True. I'll change to span<const Shape*>